pandas-stubs/core/groupby/generic.pyi

```diff
@@ -111,7 +111,13 @@ class _DataFrameGroupByNonScalar(DataFrameGroupBy):
 class DataFrameGroupBy(GroupBy):
     def any(self, skipna: bool = ...) -> DataFrame: ...
     def all(self, skipna: bool = ...) -> DataFrame: ...
-    def apply(self, func, *args, **kwargs) -> DataFrame: ...
+    # mypy sees the two overloads as overlapping
+    @overload
+    def apply(  # type: ignore[misc]
+        self, func: Callable[[DataFrame], Series | Scalar], *args, **kwargs
+    ) -> Series: ...
+    @overload
+    def apply(self, func: Callable, *args, **kwargs) -> DataFrame: ...
```

Comment:
They are overlapping `Callable[[DataFrame], Series | Scalar]` is also `Callable` but their return type is different.

Not sure whether it is possible, you can add more specialized functions for the DataFrame case (so that the first and second overload do not overlap anymore). And at the end have a "catch-all" overload `def apply(self, func: Callable, *args, **kwargs) -> DataFrame | Series: ...`

Reply:
`pyright` doesn't complain about this - just `mypy` does.

Just did a commit where I added another `overlay` and made the catchall return `DataFrame | Series` as you suggested.